Proof serialization

Not sure if I'm doing it correctly but it seems there's an issue with serializing the proofs (compressed supernova to be exact). 

error[E0277]: the trait bound `MultiFrame<'_, Fq, Sha256Coproc<Fq>>: serde::ser::Serialize` is not satisfied
   --> examples/sha256_nivc.rs:121:34
    |
121 |     let buf = bincode::serialize(&compressed_proof);
    |               ------------------ ^^^^^^^^^^^^^^^^^ the trait `serde::ser::Serialize` is not implemented for `MultiFrame<'_, Fq, Sha256Coproc<Fq>>`

If I go into the sha256.rs file and add this it works:

```
impl<'a, Fq> Serialize for MultiFrame<'a, Fq, Sha256Coproc<Fq>> where Fq: Serialize + LurkField {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where
            S: Serializer,
    {
        serializer.serialize_unit()
    }
}
```

However, this does not work if you define your custom coprocessor outside lurk-rs as you can't create an implementation for something outside of the crate. 

Is there any way to work around this?

I only added these two lines to sha256_nivc.rs after the compression:
```
let buf = bincode::serialize(&compressed_proof);
println!("{:?}", buf.unwrap().len());
```

Everything else is the same:

The full error is:
```
error[E0277]: the trait bound `MultiFrame<'_, Fq, Sha256Coproc<Fq>>: serde::ser::Serialize` is not satisfied
   --> examples/sha256_nivc.rs:125:34
    |
125 |     let buf = bincode::serialize(&compressed_proof);
    |               ------------------ ^^^^^^^^^^^^^^^^^ the trait `serde::ser::Serialize` is not implemented for `MultiFrame<'_, Fq, Sha256Coproc<Fq>>`
    |               |
    |               required by a bound introduced by this call
    |
    = help: the following other types implement trait `serde::ser::Serialize`:
              bool
              char
              isize
              i8
              i16
              i32
              i64
              i128
            and 305 others
    = note: required for `lurk::proof::supernova::Proof<'_, Fq, Sha256Coproc<Fq>, MultiFrame<'_, Fq, Sha256Coproc<Fq>>>` to implement `serde::ser::Serialize`
note: required by a bound in `bincode::serialize`
   --> /home/chris/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bincode-1.3.3/src/lib.rs:108:8
    |
106 | pub fn serialize<T: ?Sized>(value: &T) -> Result<Vec<u8>>
    |        --------- required by a bound in this function
107 | where
108 |     T: serde::Serialize,
    |        ^^^^^^^^^^^^^^^^ required by this bound in `serialize`
```

I've tried adding an annotation to arecibo:: snark::CompressedSNARK (and updating the cargo.toml to point at my local copy). But I still get the same error.

```
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(bound = "")]
pub struct CompressedSNARK<E1, E2, C1, C2, S1, S2>
where
  E1: Engine<Base = <E2 as Engine>::Scalar>,
  E2: Engine<Base = <E1 as Engine>::Scalar>,
  C1: StepCircuit<E1::Scalar>,
  C2: StepCircuit<E2::Scalar>,
  S1: BatchedRelaxedR1CSSNARKTrait<E1>,
  S2: RelaxedR1CSSNARKTrait<E2>,
{
  r_U_primary: Vec<RelaxedR1CSInstance<E1>>,
  r_W_snark_primary: S1,

  r_U_secondary: RelaxedR1CSInstance<E2>,
  l_u_secondary: R1CSInstance<E2>,
  nifs_secondary: NIFS<E2>,
  f_W_snark_secondary: S2,

  num_steps: usize,
  program_counter: E1::Scalar,

  zn_primary: Vec<E1::Scalar>,
  zn_secondary: Vec<E2::Scalar>,

  #[serde(skip)]
  _p: PhantomData<(E1, E2, C1, C2, S1, S2)>,
}
```

--- a/examples/sha256_ivc.rs
+++ b/examples/sha256_ivc.rs
@@ -112,6 +112,9 @@ fn main() {
 
     println!("Compression took {:?}", compress_end);
 
+    let buf = bincode::serialize(&compressed_proof).unwrap();
+    println!("proof size : {:}B", buf.len());
+
     let compressed_verify_start = Instant::now();
     let res = compressed_proof.verify(&pp, &z0, &zi).unwrap();
     let compressed_verify_end = compressed_verify_start.elapsed();

--- a/examples/sha256_nivc.rs
+++ b/examples/sha256_nivc.rs
@@ -116,6 +116,9 @@ fn main() {
 
     println!("Compression took {:?}", compress_end);
 
+    let buf = bincode::serialize(&compressed_proof).unwrap();
+    println!("proof size : {:}B", buf.len());
+
     let compressed_verify_start = Instant::now();
     let res = compressed_proof.verify(&pp, &z0, &zi).unwrap();
     let compressed_verify_end = compressed_verify_start.elapsed();

--- a/src/proof/supernova.rs
+++ b/src/proof/supernova.rs
@@ -127,6 +127,7 @@ where
 
 /// An enum representing the two types of proofs that can be generated and verified.
 #[derive(Serialize, Deserialize)]
+#[serde(bound = "")]
 pub enum Proof<
     'a,
     F: CurveCycleEquipped,

